src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -140,57 +114,65 @@ uct_rdmacm_cm_query_reserved_qpn_cap_2(struct ibv_context *verbs,
         goto err_out;
     }
 
-    ctx->reserved_qpn_support                          = true;
-    ctx->reserved_qpn_ctx.log_reserved_qpn_granularity =
+    ctx->log_reserved_qpn_granularity =
             UCT_IB_MLX5DV_GET(cmd_hca_cap_2, cap, log_reserved_qpn_granularity);
     ucs_debug("reserved qpn cap: log_reserved_qpn_granularity is 0x%x",
-              ctx->reserved_qpn_ctx.log_reserved_qpn_granularity);
+              ctx->log_reserved_qpn_granularity);
+
+    ctx->reserved_qpn_support = 1;
+
+    ucs_spinlock_init(&ctx->lock, 0);
+    ucs_list_head_init(&ctx->blk_list);
     return UCS_OK;
 
 err_out:
 #endif
-    ctx->reserved_qpn_support = false;
-    return UCS_ERR_UNSUPPORTED;
+    ctx->reserved_qpn_support = 0;
+    /* Create a dummy completion queue */
+    ctx->cq = ibv_create_cq(verbs, 1, NULL, NULL, 0);
+    if (ctx->cq == NULL) {
+        ucs_error("ibv_create_cq() failed: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
 }
 
-static ucs_status_t
-uct_rdmacm_cm_reserved_qpn_blk_list_free(struct uct_rdmacm_cm_qp_context *ctx)
+static void
+uct_rdmacm_cm_device_context_cleanup(struct uct_rdmacm_cm_device_context *ctx)
 {
-    ucs_status_t status = UCS_OK;
-
 #if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    ucs_status_t status;
     struct uct_rdmacm_cm_reserved_qpn_blk *blk, *tmp;
 
-    pthread_mutex_lock(&ctx->reserved_qpn_ctx.mutex);
+    ucs_spin_lock(&ctx->lock);
 
-    ucs_list_for_each_safe(blk, tmp, &ctx->reserved_qpn_ctx.blk_list, entry) {
-        ucs_list_del(&blk->entry);
-        status = uct_rdmacm_cm_reserved_qpn_blk_free(blk);
+    ucs_list_for_each_safe(blk, tmp, &ctx->blk_list, entry) {
+        status = uct_rdmacm_cm_reserved_qpn_blk_destroy(blk);
         if (status != UCS_OK) {
-            ucs_error("uct_rdmacm_cm_reserved_qpn_blk_list_free failed");
+            ucs_warn("uct_rdmacm_cm_device_context_cleanup failed");
         }
     }
 
-    pthread_mutex_unlock(&ctx->reserved_qpn_ctx.mutex);
+    ucs_spin_unlock(&ctx->lock);
 #endif
 
-    return status;
+    return;
 }
 
-static void uct_rdmacm_cm_qp_context_cleanup(uct_rdmacm_cm_t *cm)
+static void uct_rdmacm_cm_cleanup_devices(uct_rdmacm_cm_t *cm)
 {
-    struct uct_rdmacm_cm_qp_context *ctx;
-    ucs_status_t status;
+    struct uct_rdmacm_cm_device_context *ctx;
     int ret;
 
     kh_foreach_value(&cm->ctxs, ctx, {
         if (ctx->reserved_qpn_support) {
             /* There can be some blks are not fully used, then they won't be 
                destroyed in RDMACM CM EP, so need to be destroyed here. */
-            status = uct_rdmacm_cm_reserved_qpn_blk_list_free(ctx);
-            ucs_assert_always(status == UCS_OK);
+            uct_rdmacm_cm_device_context_cleanup(ctx);
+            ucs_spinlock_destroy(&ctx->lock);
         } else {
-            ret = ibv_destroy_cq(ctx->dummy_qp_ctx.cq);
+            ret = ibv_destroy_cq(ctx->cq);
```

Comment:
pls move ibv_destroy_cq code into uct_rdmacm_cm_device_context_cleanup, so it will be on par with uct_rdmacm_cm_device_context_init